### PR TITLE
SEMI baseline: merge benefit and group&ranking

### PR DESCRIPTION
**Pull request by KatGarmash**
* assumed API: takes list of oportunities (line ranges) and dict from line number to list referenced vars

### CODE REVIEW

File: veniq/baselines/semi/group_and_rank.py [hunk 1]
```diff
@@ -69,15 +55,18 @@ def group_and_rank_in_groups(
...
     """
-    primary_oportunities = group_and_rank_in_groups(oportunities, **kwargs)
-    primary_oportunities_eval = {o: _temp_fitness_f(o) for o in
+    primary_oportunities = group_and_rank_in_groups(line_to_semantic_dict,
+                                                    opportunities_struct,
+                                                    **kwargs)
+    primary_oportunities_eval = {o: _get_benefit(line_to_semantic_dict, o) for o in
                                  primary_oportunities}
     sorted_oport = sorted(primary_oportunities_eval.items(),
                           key=lambda x: x[1], reverse=True)
```
Comment (acheshkov) [inline veniq/baselines/semi/group_and_rank.py:72]:
```suggestion
    sorted_oport = sorted(
        primary_oportunities_eval.items(),
        key=lambda x: x[1],
        reverse=True
    )

```

File: veniq/baselines/semi/benefits.py [hunk 2]
```diff
@@ -1,81 +1,53 @@
...
-        # method_name = method_ast.get_root().name
-        original_method_semantic = extract_method_statements_semantic(method_ast)
-    reprocessed_dict = _reprocess_dict(original_method_semantic)
-    return reprocessed_dict
-
-
-def _LCOM2(file_dict: Dict, range_stats=[], mode='original') -> int:
+def _LCOM2(line_to_semantic_dict: Dict,
+           range_statements: Tuple[int, int] = None,
+           mode='original') -> int:
```
Comment (acheshkov) [inline veniq/baselines/semi/benefits.py:23]:
```suggestion
def _LCOM2(
    line_to_semantic_dict: Dict,
    range_statements: Tuple[int, int] = None,
    mode='original'
) -> int:
```

File: test/baselines/semi/test_group_and_rank.py [hunk 3]
```diff
@@ -37,35 +61,16 @@ def test_not_overlap_custom(self):
...
-    def test_output_best_oportunities_custom_overlap_top4(self):
-        expect_list = [self.oport_0, self.oport_1,
-                       self.oport_3, self.oport_4]
-        self.assertEqual(output_best_oportunities(self.oportunities,
-                                                  top_k=4,
-                                                  min_overlap=0.9),
-                         expect_list)
+        selected_primary = group_and_rank_in_groups(self.line_to_semantic_dict,
+                                                    self.opportunities,
+                                                    max_size_difference=0.25)
```
Comment (acheshkov) [inline test/baselines/semi/test_group_and_rank.py:66]:
```suggestion
        selected_primary = group_and_rank_in_groups(
            self.line_to_semantic_dict,
            self.opportunities,
            max_size_difference=0.25
        )

```

File: veniq/baselines/semi/benefits.py [hunk 4]
```diff
@@ -1,81 +1,53 @@
...
+        statement_2: int
+    ) -> bool:
     '''
     This function is aimed to check whether 2 statements have
     common semantics or not.
     '''
-    joined_names: Counter = Counter(dict_file[statement_1] + dict_file[statement_2])
+
+    joined_names: Counter = Counter(defaultdict(lambda: [], dict_file)[statement_1] +
+                                    defaultdict(lambda: [], dict_file)[statement_2])
```
Comment (acheshkov) [inline veniq/baselines/semi/benefits.py:16]:
```suggestion
    joined_names: Counter = Counter(
        defaultdict(lambda: [], dict_file)[statement_1] +
        defaultdict(lambda: [], dict_file)[statement_2]
    )
```

File: veniq/baselines/semi/benefits.py [hunk 5]
```diff
@@ -1,81 +1,53 @@
...
+    line_to_semantic_dict is a map from line number
+    to variables that are referenced there.
+    range_statements is the range of lines for which LCOM2
+    is calculated.
     '''
     P = 0
     Q = 0
     list_statements = []
 
     if mode == 'after_ref':
```
Comment (acheshkov) [inline veniq/baselines/semi/benefits.py:39]: what is `after_ref`? Consider creating ENUM structure

**Review by acheshkov — COMMENTED** [review]
Consider the creation of class for `Ranges`.

File: veniq/baselines/semi/benefits.py (hunk 5 above)
Comment (KatGarmash) [reply veniq/baselines/semi/benefits.py:43]: it's @Vitaly-Protasov 's code
it's an indicator meaning that we consider the source code except for the range selected for extraction

We can refactor it next week

**Review by acheshkov — APPROVED** [review]

**Pull request merged**